program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,375 @@ use crate::c_oracle_header::{
+        }
+    }
+}
+
+
+impl<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64>
+    Tracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+    for TickTracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+{
+    fn initialize(&mut self) -> Result<(), OracleError> {
+        Ok(())
+    }
 
+    ///add a new price to the tracker
+    fn add_price(
+        &mut self,
+        prev_time: i64,
+        prev_price: i64,
+        prev_conf: u64,
+        current_time: i64,
+        current_price: i64,
+        current_conf: u64,
+    ) -> Result<(), OracleError> {
+        let prev_entry = Self::time_to_entry(prev_time)?;
+        let current_entry = Self::time_to_entry(current_time)?;
+        let num_skipped_entries = current_entry - prev_entry;
+        match num_skipped_entries {
+            0 => {
+                self.prices[current_entry] = current_price;
+                self.confidences[current_entry] = current_conf;
+            }
+            1 => {
+                self.prices[prev_entry] = prev_price;
+                self.confidences[prev_entry] = prev_conf;
+                self.entry_validity[prev_entry] =
+                    if Self::get_time_to_entry_end(prev_time) >= THRESHOLD {
+                        0
+                    } else {
+                        1
+                    };
+                self.prices[current_entry] = current_price;
+                self.confidences[current_entry] = current_conf;
+            }
+            _ => {
+                //invalidate all the entries in your way
+                //this is ok because THRESHOLD < Granuality
+                self.prices[prev_entry] = prev_price;
+                self.confidences[prev_entry] = prev_conf;
+                self.entry_validity[prev_entry] =
+                    if Self::get_time_to_entry_end(prev_time) >= THRESHOLD {
+                        0
+                    } else {
+                        1
+                    };
+
+                self.invaldate_following_entries(
+                    cmp::min(num_skipped_entries - 1, NUM_ENTRIES),
+                    Self::get_next_entry(prev_entry),
+                );
+                self.prices[current_entry] = current_price;
+                self.confidences[current_entry] = current_conf;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+const THIRTY_MINUTES: i64 = 30 * 60;
+const TEN_HOURS: usize = 10 * 60 * 60;
+const TWENTY_SECONDS: i64 = 20;
 #[derive(Debug, Clone, Copy)]
 #[repr(C)]
 /// this wraps multiple SMA and tick trackers, and includes all the state
 /// used by the time machine
 pub struct TimeMachineWrapper {
-    //Place holder with the size of the fields I am planning to add
-    place_holder: [u8; 1864],
+    sma_tracker:
+        SmaTracker<THIRTY_MINUTES, { TEN_HOURS / (THIRTY_MINUTES as usize) }, TWENTY_SECONDS>,
+    tick_tracker:
+        TickTracker<THIRTY_MINUTES, { TEN_HOURS / (THIRTY_MINUTES as usize) }, TWENTY_SECONDS>,
```

Comment:
maybe we should do 24 hours instead of 10? that feels a little less arbitrary (because it's a whole day). we can then also roll up to longer durations, like we store 24 hour SMAs for the past 7 days.

Also, i think all of these quantities need to be measured in milliseconds (which i believe is the resolution of the timestamps)

Reply:
i think we need to discuss what windows what we want to provide, I just left this is a placeholder

Reply:
makes sense. fwiw you can head off these kinds of comments by putting a `// TODO: ` comment in the code. often a good trick for answering questions that you know reviewers will ask.